README.md
# UIT AI Challenge - Artistic Text Challenge

## Dependencies
Tiến hành cài đặt các dependencies trước có thể chạy
```
pip install -r requirements.txt
```

## Xử lý dữ liệu
### Đối với dữ liệu để huấn luyện cho model Recognition

1. Đầu tiên ta tiến hành tải bộ dữ liệu đã được đánh nhãn từ BTC [tại đây](https://drive.google.com/file/u/3/d/1NJJA1A8I2Xj5-107E3DFohBNzjWyaaf7/view?usp=share_link)
2. Giải nén và thay đổi tên folder thành `data` và copy vào thư mục `tools/Recognition` với format là:
```
data
├── images 		- chứa các ảnh
├── labels 		- chứa các file json
```
3. Thực thi câu lệnh `./process_rec.sh` để tiến hành xử lý các ảnh, sinh thêm các synthetics data, augment data và split data thành tập dữ liệu train và val (tỉ lệ 80-20)

> Team có thực hiện thêm việc kiếm tra các ảnh bị đánh nhãn sai, nên số lượng ảnh và quá trình thực hiện có khác đôi chút với hướng dẫn bên trên.


PaddleOCR/history
    1  curl -L git.io/antigen > antigen.zsh
    2  mv antigen.zsh .antigen.zsh
    3  rm -r .oh-my-zsh
    4  apt install neovim
    5  asdf
    6  apt-get install software-properties-common
    7  sudo add-apt-repository ppa:neovim-ppa/stable
    8  sudo apt-get install neovim
    9  mkdir -p ~/.config/nvim
   10  ls /root/
   11  vi Downloads
   12  unlink Downloads
   13  find .
   14  mkdir Downloads
   15  ls /home
   16  unlink Desktop
   17  unlink Documents
   18  unlink Music
   19  unlink Pictures
   20  unlink Public
   21  unlink Templates
   22  unlink Videos
   23  mkdir Desktop
   24  mkdir Document
   25  mkdir Music
   26  mkdir Pictures
   27  mkdir Public
   28  mkdir Videos
   29  mv -r ~/.config/nvim ~/
   30  mv ~/.config/nvim ~/
   31  ls -
   32  mv nvim .config
   33  ls ~/
   34  cd ~Down
   35  https://github.com/tgbaodeeptry/dotfiles.git
   36  cp -r dotfiles/.config/nvim/ ~/.config/nvim
   37  cp -r nvim .
   38  cp -r nvim/* .
   39  nvim lua/module/base.lua
   40  nvim lua/module/theme.lua
   41  nvim lua/plugins/treesitter.lua
   42  nvim init.lua
   43  nvim lua/plugins/barbar.lua
   44  nvim lua/plugins/bufferline.lua
   45  nvim lua/plugins/lualine.lua
   46  gcc -v
   47  nvim lua/plugins/neoscroll.lua
   48  nvim lua/module/mappings.lua
   49  nvim lua/plugins/lspsaga.lua
   50  nvim lua/plugins/lspconfig.lua
   51  nvim lua/plugins/lspkind.lua
   52  nvim lua/plugins/autosave.lua
   53  nvim lua/module/plugins.lua
   54  mv autosave.lua auto-save.lua
   55  vi lspsaga.lua
   56  mv lspsaga.lua lspsaga.lua.bak
   57  rm lspsaga.lua.bak
   58  mv lspsage.lua lspsaga.lua
   59  nvim lspsage.lua
   60  nvim lspconfig.lua 
   61  nvim auto-save.lua
   62  ../module/plugins.lua
   63  nvim ../module/plugins.lua
   64  nvim ../module/theme.lua
   65  which zsh
   66  nvim ../module/base.lua
   67  nvim ../plugins/auto-save.lua
   68  nvim ../plugins/bufferline.lua
   69  clera
   70  vi ../module/plugins.lua
   71  sudo apt-get install ruby-full build-essential zlib1g-dev
   72  gem install jekyll bundler
   73  snap install libvips
   74  sudo apt install snapd
   75  sudo snap install libvips
   76  systemctl status snapd.service
   77  snap --version
   78  snap version
   79  sudo apt install libvips libvips-dev
   80  cd LinuxTus
   81  bundler install
   82  apt install nodejs
   83  JEKYLL_ENV=production bundle exec jekyll build
   84  apt install pip3
   85  apt install python3-pip
   86  pip3 install requests schedule
   87  pip3 install beautifulsoup
   88  pip3 install beautifulsoup4
   89  pip3 install html5lib
   90  ma
   91  antigen version
   92  zsh version
   93  zsh --version
   94  vi ~/.antigen.zsh
   95  cd /m
   96  find . -name "08-29:
   97  find . -name "08-29"
   98  find . -name "*08-29"
   99  find . -name "*08-29*"
  100  find . -name "*08-29*" -exec rm {} \;
  101  ls ubuntu/_posts | grep "8-29"
  102  cd assets/img/posts
  103  while 1; do echo data(); done
  104  while 1; do echo date(); done
  105  while 1; do echo date() done
  106  while\n:
  107  git status -uno
  108  git push origin -f gh-pages
  109  curl -X POST "https://api-gateway.ezoic.com/gateway/cdnservices/bulkclearcache?developerKey=ae97192eac5889f8efec8a3979503a9d" -H "Content-Type: application/json" --data { "urls":[ "https://linuxtus.com/", "https://linuxtus.com", "https://linuxtus.com/sitemap.xml" ] }'\n
  110  curl -X POST "https://api-gateway.ezoic.com/gateway/cdnservices/bulkclearcache?developerKey=ae97192eac5889f8efec8a3979503a9d" -H "Content-Type: application/json" --data '{ "urls":[ "https://linuxtus.com/", "https://linuxtus.com", "https://linuxtus.com/sitemap.xml" ] }'\n
  111  cd /f
  112  mkdir C
  113  mkdir D
  114  ln -s D /mnt/d
  115  cd /mnt/d
  116  unlink D
  117  vi b
  118  ls D
  119  cd technology
  120  ssh-keygen -t rsa -b 2048
  121  ssh root@103.74.120.48
  122  cat .ssh/known_hosts
  123  cat /root/.ssh/id_rsa
  124  cat /root/.ssh
  125  ls /root/.ssh
  126  ssh-keygen -t rsa
  127  rm vps-48 vps-48.pub
  128  vi vps48.pub
  129  ssh root@103.74.120.48 -p 2512
  130  apt install sshpass
  131  rm vps48
  132  rm vps48.pub
  133  sshpass -
  134  sshpass -h
  135  sshpass -p 'luckyvn$$2512' ssh -o StrictHostKeyChecking=no -p 2512 root@103.74.120.48 'ls'
  136  sshpass -p 'luckyvn$$2512' ssh -o StrictHostKeyChecking=no -p 2512 root@103.74.120.48 
  137  sshpass -p 'luckyvn$$2512' ssh -o StrictHostKeyChecking=no -p 2512 root@103.74.120.48 'clear'
  138  uptime
  139  vi config/last_run
  140  ls b
  141  rm -r a b
  142  git clone ht
  143  git clone https://github.com/tgbaodeeptry/dotfiles.git
  144  rm -r dotfiles
  145  git clone https://github.com/tgbaodeeptry/dotfiles.git\n: 1661784776:0;git clone https://github.com/tgbaodeeptry/dotfiles.git
  146  cd ~/.config/nvim/lua/module/b
  147  vi ~/.config/nvim/lua/module/base.lua
  148  vi ~/.config/nvim/lua/module/plugins.lua
  149  vi lua/module/theme.lua
  150  vi lua/plugins/lualine.lua
  151  cd ~/.config/nvim/lua/module/
  152  vi base.lua
  153  cd plugins/
  154  vi bufferline.lua
  155  vi module/plugins.lua
  156  vi plugins/lualine.lua
  157  atom _config.yml
  158  atom.cmd _config.yml
  159  explorer.exe
  160  vi plugins/bufferline.lua
  161  mkdir /mnt/d/LinuxTusNew
  162  scp -h
  163  scp -P 25734 root@14.225.253.225:/root/Downloads/LinuxTus /mnt/d/LinuxTusNew
  164  scp -r -P 25734 root@14.225.253.225:/root/Downloads/LinuxTus /mnt/d/LinuxTusNew
  165  mkdir LinuxTusNew
  166  scp -r -P 25734 root@14.225.253.225:/root/Downloads/LinuxTus.zip /mnt/d/LinuxTusNew
  167  cd LinuxTusNew
  168  unzip LinuxTus.zip
  169  rm -r LinuxTus.zip
  170  mv LinuxTus/ LinuxTusNew
  171  cp -r LinuxTusNew/* .
  172  rm -r LinuxTusNew
  173  jekyll serve
  174  cd ubuntu/_posts
  175  rm -r 2022-08-30-*
  176  rm -r 2022-08-29-*
  177  rm -r 2022-08-28-*
  178  rm -r 2022-08-27-*
  179  rm -r 2022-08-26-*
  180  rm -r 2022-08-25-*
  181  rm -r 2022-08-24-*
  182  rm -r 2022-08-23-*

YOLOv7/test2.py
import argparse
import json
import os
from pathlib import Path
from threading import Thread

import numpy as np
import torch
import yaml
from tqdm import tqdm

from models.experimental import attempt_load
from utils.datasets import create_dataloader
from utils.general import coco80_to_coco91_class, check_dataset, check_file, check_img_size, box_iou, \
    non_max_suppression, scale_coords, xyxy2xywh, xywh2xyxy, set_logging, increment_path, colorstr
from utils.metrics import ap_per_class, ConfusionMatrix
from utils.plots import plot_images, output_to_target, plot_study_txt
from utils.torch_utils import select_device, time_synchronized, TracedModel


def test(data,
         weights=None,
         batch_size=32,
         imgsz=640,
         conf_thres=0.001,
         iou_thres=0.6,  # for NMS
         save_json=False,
         single_cls=False,
         augment=False,
         verbose=False,
         model=None,
         dataloader=None,
         save_dir=Path(''),  # for saving images
         save_txt=False,  # for auto-labelling
         save_hybrid=False,  # for hybrid auto-labelling
         save_conf=False,  # save auto-label confidences
         plots=True,
         wandb_logger=None,
         compute_loss=None,
         half_precision=True,
         trace=False,
         is_coco=False,
         v5_metric=False):
    # Initialize/load model and set device
    training = model is not None
    if training:  # called by train.py
        device = next(model.parameters()).device  # get model device

    else:  # called directly
        set_logging()
        device = select_device(opt.device, batch_size=batch_size)

        # Directories
        save_dir = Path(increment_path(Path(opt.project) / opt.name, exist_ok=opt.exist_ok))  # increment run
        (save_dir / 'labels' if save_txt else save_dir).mkdir(parents=True, exist_ok=True)  # make dir

        # Load model
        model = attempt_load(weights, map_location=device)  # load FP32 model
        gs = max(int(model.stride.max()), 32)  # grid size (max stride)
        imgsz = check_img_size(imgsz, s=gs)  # check img_size

        if trace:
            model = TracedModel(model, device, imgsz)

    # Half
    half = device.type != 'cpu' and half_precision  # half precision only supported on CUDA
    if half:
        model.half()

    # Configure
    model.eval()
    if isinstance(data, str):
        is_coco = data.endswith('coco.yaml')
        with open(data) as f:
            data = yaml.load(f, Loader=yaml.SafeLoader)
    check_dataset(data)  # check
    nc = 1 if single_cls else int(data['nc'])  # number of classes
    # iouv = torch.linspace(0.5, 0.95, 10).to(device)  # iou vector for mAP@0.5:0.95
    iouv = torch.linspace(0.5, 0.5, 1).to(device)  # iou vector for mAP@0.5:0.95
    niou = iouv.numel()

    # Logging
    log_imgs = 0
    if wandb_logger and wandb_logger.wandb:
        log_imgs = min(wandb_logger.log_imgs, 100)
    # Dataloader
    if not training:
        if device.type != 'cpu':
            model(torch.zeros(1, 3, imgsz, imgsz).to(device).type_as(next(model.parameters())))  # run once
        task = opt.task if opt.task in ('train', 'val', 'test') else 'val'  # path to train/val/test images
        dataloader = create_dataloader(data[task], imgsz, batch_size, gs, opt, pad=0.5, rect=True,
                                       prefix=colorstr(f'{task}: '))[0]

    if v5_metric:
        print("Testing with YOLOv5 AP metric...")

    seen = 0
    confusion_matrix = ConfusionMatrix(nc=nc, conf=conf_thres)
    names = {k: v for k, v in enumerate(model.names if hasattr(model, 'names') else model.module.names)}
    coco91class = coco80_to_coco91_class()
    s = ('%20s' + '%12s' * 6) % ('Class', 'Images', 'Labels', 'P', 'R', 'mAP@.5', 'mAP@.5:.95')
    p, r, f1, mp, mr, map50, map, t0, t1 = 0., 0., 0., 0., 0., 0., 0., 0., 0.
    loss = torch.zeros(3, device=device)
    jdict, stats, ap, ap_class, wandb_images = [], [], [], [], []
    for batch_i, (img, targets, paths, shapes) in enumerate(tqdm(dataloader, desc=s)):
        img = img.to(device, non_blocking=True)
        img = img.half() if half else img.float()  # uint8 to fp16/32
        img /= 255.0  # 0 - 255 to 0.0 - 1.0
        targets = targets.to(device)
        nb, _, height, width = img.shape  # batch size, channels, height, width

        with torch.no_grad():
            # Run model
            t = time_synchronized()
            out, train_out = model(img, augment=augment)  # inference and training outputs
            t0 += time_synchronized() - t

            # Compute loss
            if compute_loss:
                loss += compute_loss([x.float() for x in train_out], targets)[1][:3]  # box, obj, cls

            # Run NMS
            targets[:, 2:] *= torch.Tensor([width, height, width, height]).to(device)  # to pixels
            lb = [targets[targets[:, 0] == i, 1:] for i in range(nb)] if save_hybrid else []  # for autolabelling
            t = time_synchronized()
            out = non_max_suppression(out, conf_thres=conf_thres, iou_thres=iou_thres, labels=lb, multi_label=True)
            t1 += time_synchronized() - t

        # Statistics per image
        for si, pred in enumerate(out):
            labels = targets[targets[:, 0] == si, 1:]
            nl = len(labels)
            tcls = labels[:, 0].tolist() if nl else []  # target class
            path = Path(paths[si])
            seen += 1

            if len(pred) == 0:
                if nl:
                    stats.append((torch.zeros(0, niou, dtype=torch.bool), torch.Tensor(), torch.Tensor(), tcls))
                continue

            # Predictions
            predn = pred.clone()
            scale_coords(img[si].shape[1:], predn[:, :4], shapes[si][0], shapes[si][1])  # native-space pred

            # Append to text file
            if save_txt:
                gn = torch.tensor(shapes[si][0])[[1, 0, 1, 0]]  # normalization gain whwh
                for *xyxy, conf, cls in predn.tolist():
                    xywh = (xyxy2xywh(torch.tensor(xyxy).view(1, 4)) / gn).view(-1).tolist()  # normalized xywh
                    line = (cls, *xywh, conf) if save_conf else (cls, *xywh)  # label format
                    with open(save_dir / 'labels' / (path.stem + '.txt'), 'a') as f:
                        f.write(('%g ' * len(line)).rstrip() % line + '\n')

            # W&B logging - Media Panel Plots
            if len(wandb_images) < log_imgs and wandb_logger.current_epoch > 0:  # Check for test operation
                if wandb_logger.current_epoch % wandb_logger.bbox_interval == 0:
                    box_data = [{"position": {"minX": xyxy[0], "minY": xyxy[1], "maxX": xyxy[2], "maxY": xyxy[3]},
                                 "class_id": int(cls),
                                 "box_caption": "%s %.3f" % (names[cls], conf),
                                 "scores": {"class_score": conf},
                                 "domain": "pixel"} for *xyxy, conf, cls in pred.tolist()]
                    boxes = {"predictions": {"box_data": box_data, "class_labels": names}}  # inference-space
                    wandb_images.append(wandb_logger.wandb.Image(img[si], boxes=boxes, caption=path.name))
            wandb_logger.log_training_progress(predn, path, names) if wandb_logger and wandb_logger.wandb_run else None

            # Append to pycocotools JSON dictionary
            if save_json:
                # [{"image_id": 42, "category_id": 18, "bbox": [258.15, 41.29, 348.26, 243.78], "score": 0.236}, ...
                image_id = int(path.stem) if path.stem.isnumeric() else path.stem
                box = xyxy2xywh(predn[:, :4])  # xywh
                box[:, :2] -= box[:, 2:] / 2  # xy center to top-left corner
                for p, b in zip(pred.tolist(), box.tolist()):
                    jdict.append({'image_id': image_id,
                                  'category_id': coco91class[int(p[5])] if is_coco else int(p[5]),
                                  'bbox': [round(x, 3) for x in b],
                                  'score': round(p[4], 5)})

            # Assign all predictions as incorrect
            correct = torch.zeros(pred.shape[0], niou, dtype=torch.bool, device=device)
            if nl:
                detected = []  # target indices
                tcls_tensor = labels[:, 0]

                # target boxes
                tbox = xywh2xyxy(labels[:, 1:5])
                scale_coords(img[si].shape[1:], tbox, shapes[si][0], shapes[si][1])  # native-space labels
                if plots:
                    confusion_matrix.process_batch(predn, torch.cat((labels[:, 0:1], tbox), 1))

                # Per target class
                for cls in torch.unique(tcls_tensor):
                    ti = (cls == tcls_tensor).nonzero(as_tuple=False).view(-1)  # prediction indices
                    pi = (cls == pred[:, 5]).nonzero(as_tuple=False).view(-1)  # target indices

                    # Search for detections
                    if pi.shape[0]:
                        # Prediction to target ious
                        ious, i = box_iou(predn[pi, :4], tbox[ti]).max(1)  # best ious, indices

                        # Append detections
                        detected_set = set()
                        for j in (ious > iouv[0]).nonzero(as_tuple=False):
                            d = ti[i[j]]  # detected target
                            if d.item() not in detected_set:
                                detected_set.add(d.item())
                                detected.append(d)
                                correct[pi[j]] = ious[j] > iouv  # iou_thres is 1xn
                                if len(detected) == nl:  # all targets already located in image
                                    break

            # Append statistics (correct, conf, pcls, tcls)
            stats.append((correct.cpu(), pred[:, 4].cpu(), pred[:, 5].cpu(), tcls))

        # Plot images
        if plots and batch_i < 3:
            f = save_dir / f'test_batch{batch_i}_labels.jpg'  # labels
            Thread(target=plot_images, args=(img, targets, paths, f, names), daemon=True).start()
            f = save_dir / f'test_batch{batch_i}_pred.jpg'  # predictions
            Thread(target=plot_images, args=(img, output_to_target(out), paths, f, names), daemon=True).start()

    # Compute statistics
    stats = [np.concatenate(x, 0) for x in zip(*stats)]  # to numpy
    print(stats)
    if len(stats) and stats[0].any():
        p, r, ap, f1, ap_class = ap_per_class(*stats, plot=plots, v5_metric=v5_metric, save_dir=save_dir, names=names)
        ap50, ap = ap[:, 0], ap.mean(1)  # AP@0.5, AP@0.5:0.95
        mp, mr, map50, map = p.mean(), r.mean(), ap50.mean(), ap.mean()
        nt = np.bincount(stats[3].astype(np.int64), minlength=nc)  # number of targets per class
    else:
        nt = torch.zeros(1)

    # Print results
    pf = '%20s' + '%12i' * 2 + '%12.3g' * 4  # print format
    print(pf % ('all', seen, nt.sum(), mp, mr, map50, map))

    # Print results per class
    if (verbose or (nc < 50 and not training)) and nc > 1 and len(stats):
        for i, c in enumerate(ap_class):
            print(pf % (names[c], seen, nt[c], p[i], r[i], ap50[i], ap[i]))

    # Print speeds
    t = tuple(x / seen * 1E3 for x in (t0, t1, t0 + t1)) + (imgsz, imgsz, batch_size)  # tuple
    if not training:
        print('Speed: %.1f/%.1f/%.1f ms inference/NMS/total per %gx%g image at batch-size %g' % t)

    # Plots
    if plots:
        confusion_matrix.plot(save_dir=save_dir, names=list(names.values()))
        if wandb_logger and wandb_logger.wandb:
            val_batches = [wandb_logger.wandb.Image(str(f), caption=f.name) for f in sorted(save_dir.glob('test*.jpg'))]
            wandb_logger.log({"Validation": val_batches})
    if wandb_images:
        wandb_logger.log({"Bounding Box Debugger/Images": wandb_images})

    # Save JSON
    if save_json and len(jdict):
        w = Path(weights[0] if isinstance(weights, list) else weights).stem if weights is not None else ''  # weights
        anno_json = './coco/annotations/instances_val2017.json'  # annotations json
        pred_json = str(save_dir / f"{w}_predictions.json")  # predictions json
        print('\nEvaluating pycocotools mAP... saving %s...' % pred_json)
        with open(pred_json, 'w') as f:
            json.dump(jdict, f)

        try:  # https://github.com/cocodataset/cocoapi/blob/master/PythonAPI/pycocoEvalDemo.ipynb
            from pycocotools.coco import COCO
            from pycocotools.cocoeval import COCOeval

            anno = COCO(anno_json)  # init annotations api
            pred = anno.loadRes(pred_json)  # init predictions api
            eval = COCOeval(anno, pred, 'bbox')
            if is_coco:
                eval.params.imgIds = [int(Path(x).stem) for x in dataloader.dataset.img_files]  # image IDs to evaluate
            eval.evaluate()
            eval.accumulate()
            eval.summarize()
            map, map50 = eval.stats[:2]  # update results (mAP@0.5:0.95, mAP@0.5)
        except Exception as e:
            print(f'pycocotools unable to run: {e}')

    # Return results
    model.float()  # for training
    if not training:
        s = f"\n{len(list(save_dir.glob('labels/*.txt')))} labels saved to {save_dir / 'labels'}" if save_txt else ''
        print(f"Results saved to {save_dir}{s}")
    maps = np.zeros(nc) + map
    for i, c in enumerate(ap_class):
        maps[c] = ap[i]
    return (mp, mr, map50, map, *(loss.cpu() / len(dataloader)).tolist()), maps, t


if __name__ == '__main__':
    parser = argparse.ArgumentParser(prog='test.py')
    parser.add_argument('--weights', nargs='+', type=str, default='yolov7.pt', help='model.pt path(s)')
    parser.add_argument('--data', type=str, default='data/coco.yaml', help='*.data path')
    parser.add_argument('--batch-size', type=int, default=32, help='size of each image batch')
    parser.add_argument('--img-size', type=int, default=640, help='inference size (pixels)')
    parser.add_argument('--conf-thres', type=float, default=0.001, help='object confidence threshold')
    parser.add_argument('--iou-thres', type=float, default=0.65, help='IOU threshold for NMS')
    parser.add_argument('--task', default='val', help='train, val, test, speed or study')
    parser.add_argument('--device', default='', help='cuda device, i.e. 0 or 0,1,2,3 or cpu')
    parser.add_argument('--single-cls', action='store_true', help='treat as single-class dataset')
    parser.add_argument('--augment', action='store_true', help='augmented inference')
    parser.add_argument('--verbose', action='store_true', help='report mAP by class')
    parser.add_argument('--save-txt', action='store_true', help='save results to *.txt')
    parser.add_argument('--save-hybrid', action='store_true', help='save label+prediction hybrid results to *.txt')
    parser.add_argument('--save-conf', action='store_true', help='save confidences in --save-txt labels')
    parser.add_argument('--save-json', action='store_true', help='save a cocoapi-compatible JSON results file')
    parser.add_argument('--project', default='runs/test', help='save to project/name')
    parser.add_argument('--name', default='exp', help='save to project/name')
    parser.add_argument('--exist-ok', action='store_true', help='existing project/name ok, do not increment')
    parser.add_argument('--no-trace', action='store_true', help='don`t trace model')
    parser.add_argument('--v5-metric', action='store_true', help='assume maximum recall as 1.0 in AP calculation')
    opt = parser.parse_args()
    opt.save_json |= opt.data.endswith('coco.yaml')
    opt.data = check_file(opt.data)  # check file
    print(opt)
    # check_requirements()

    if opt.task in ('train', 'val', 'test'):  # run normally
        test(opt.data,
             opt.weights,
             opt.batch_size,
             opt.img_size,
             opt.conf_thres,
             opt.iou_thres,
             opt.save_json,
             opt.single_cls,
             opt.augment,
             opt.verbose,
             save_txt=opt.save_txt | opt.save_hybrid,
             save_hybrid=opt.save_hybrid,
             save_conf=opt.save_conf,
             trace=not opt.no_trace,
             v5_metric=opt.v5_metric
             )

    elif opt.task == 'speed':  # speed benchmarks
        for w in opt.weights:
            test(opt.data, w, opt.batch_size, opt.img_size, 0.25, 0.45, save_json=False, plots=False,
                 v5_metric=opt.v5_metric)

    elif opt.task == 'study':  # run over a range of settings and save/plot
        # python test.py --task study --data coco.yaml --iou 0.65 --weights yolov7.pt
        x = list(range(256, 1536 + 128, 128))  # x axis (image sizes)
        for w in opt.weights:
            f = f'study_{Path(opt.data).stem}_{Path(w).stem}.txt'  # filename to save to
            y = []  # y axis
            for i in x:  # img-size
                print(f'\nRunning {f} point {i}...')
                r, _, t = test(opt.data, w, opt.batch_size, i, opt.conf_thres, opt.iou_thres, opt.save_json,
                               plots=False, v5_metric=opt.v5_metric)
                y.append(r + t)  # results and times
            np.savetxt(f, y, fmt='%10.4g')  # save
        os.system('zip -r study.zip study_*.txt')
        plot_study_txt(x=x)  # plot
